lib/CL/devices/hsa/pocl-hsa.c

```diff
@@ -118,6 +126,112 @@ pocl_hsa_get_agents(hsa_agent_t agent, void *data)
   return HSA_STATUS_SUCCESS;
 }
 
+//Get hsa-unsupported device features from hsa device list.
+int num_hsa_device = 1;
+hsa_device_features hsa_device_lists[MAX_HSA_AGENTS];
+void init_hsa_device_list()
+{
+	hsa_device_lists[0].dev_name = "Spectre";
+	hsa_device_lists[0].llvm_cpu = "kaveri";
+	//FIXME: I am not sure if it is related to main memory.
+	hsa_device_lists[0].max_mem_alloc_size = 592969728;
+	hsa_device_lists[0].global_mem_size = 2371878912;
+	hsa_device_lists[0].vendor_id = 0x1002;
+	hsa_device_lists[0].global_mem_cache_type = 0x2;
+	hsa_device_lists[0].global_mem_cacheline_size = 64;
+	hsa_device_lists[0].max_compute_units = 8;
+	hsa_device_lists[0].max_clock_frequency = 720;
+	hsa_device_lists[0].max_conastant_buffer_size = 65536;
+	hsa_device_lists[0].local_mem_size = 32768;
+}
+
+void get_hsa_device_features(char* dev_name, struct _cl_device_id* dev)
+{
+	int i;
+	for(i=0; i<num_hsa_device; i++){
+		if(strcmp(dev_name, hsa_device_lists[i].dev_name) == 0){
+			dev->llvm_cpu = hsa_device_lists[i].llvm_cpu;
+			dev->max_mem_alloc_size = hsa_device_lists[i].max_mem_alloc_size;
+			dev->global_mem_size = hsa_device_lists[i].global_mem_size;
+			dev->vendor_id = hsa_device_lists[i].vendor_id;
+			dev->global_mem_cache_type = hsa_device_lists[i].global_mem_cache_type;
+			dev->global_mem_cacheline_size = hsa_device_lists[i].global_mem_cacheline_size;
+			dev->max_compute_units = hsa_device_lists[i].max_compute_units;
+			dev->max_clock_frequency = hsa_device_lists[i].max_clock_frequency;
+			dev->max_constant_buffer_size = hsa_device_lists[i].max_conastant_buffer_size;
+			dev->local_mem_size = hsa_device_lists[i].local_mem_size;
+			break;
+		}
+	}
+
+}
+
+void
+pocl_hsa_init_device_infos(struct _cl_device_id* dev)
+{
+  pocl_basic_init_device_infos (dev);
+  dev->spmd = CL_TRUE;
+  dev->llvm_target_triplet = "amdgcn--amdhsa";
+  dev->has_64bit_long = 1;
+  hsa_agent_t agent = hsa_agents[found_hsa_agents - 1];
+
+  hsa_status_t stat = hsa_agent_get_info(agent, HSA_AGENT_INFO_CACHE_SIZE, &(dev->global_mem_cache_size));
+  char dev_name[64] = {0};
+  stat = hsa_agent_get_info(agent, HSA_AGENT_INFO_NAME, dev_name);
```

Comment:
Hmm. This API seems quite unsafe (not your fault, of course) as it doesn't take the buffer size as an argument, thus we only have to trust the names are not longer than 64.